### DISCUSSION TITLE
Build problem

### DISCUSSION THREAD

**Discussion by jcdutton** [ccache/ccache#376] [Support]
I have had build problems on a project. If I bypass ccache it resolves the build problems. I am assuming this is due to hash collisions.
Is there any way to re-configure ccache to use a hash that maybe results in more bits in the hash, and therefore less likely to collide?

**Comment by afbjorklund** [comment]
There could be other reasons for build problems than hash collisions, it would be interesting to have more details about your issue. If it works fine without ccache, it sounds like there is indeed a problem.

Please note that ccache uses MD4 + size (=160 bits), which is more unlikely to have hash collisions than plain MD4/MD5 does.Theoretically we could move to SHA1, but there is a performance impact doing so.

**Comment by jcdutton** [comment]
Is there a command I can type that can check for collisions.
Does it store the source file that it hashed?  So, if there is a problem, one could do a source file diff and discover which one is bad?

**Comment by afbjorklund** [comment]
You can try the new debug mode, it will give some more (readable) insight in how the hash is composed.

**Comment by jrosdahl** [comment]
> I am assuming this is due to hash collisions.

Hash collisions are certainly possible but extremely unlikely, even with MD4. It's much much more likely that the problem is something else.

> Is there any way to re-configure ccache to use a hash that maybe results in more bits in the hash, and therefore less likely to collide?

Nope, MD4 is currently the only available algorithm.

> Is there a command I can type that can check for collisions.

Not really.

As @afbjorklund says, the easiest way to investigate the problem is to use the new debug mode. See https://ccache.dev/manual/latest.html#_cache_debugging for more information and hints.

**Comment by pavsa** [comment]
https://github.com/ccache/ccache/issues/374 
You might have hit the same bug as I did, it was solved only few days ago.

**Comment by jcdutton** [comment]
In my case, all the files were .d  files. So, it looks likely to have been similar to #374 
Is there any test tool to:
1) Check the cache integrity
2) Check for any collisions.

**Comment by afbjorklund** [comment]
This was a fairly new addition, not even sure if it hit any released version ?
As far as I know it only started happening after #363, no duplicates before.

**Comment by afbjorklund** [comment]
Currently the best integrity is provided by a compressed cache, courtesy of the gzip format.
As far as I know there is (yet?) no tool to check for collisions, and I think these were a bug...

**Comment by jrosdahl** [comment]
> Is there any test tool to: [...] Check the cache integrity

Nope.

> Is there any test tool to: [...] Check for any collisions.

Were you not happy with my previous answer to this question?

**Comment by jrosdahl** [comment]
> In my case, all the files were .d files. So, it looks likely to have been similar to #374

Just to make things clear: Do you use the new (opt-in) depend mode with ccache compiled from master, not from any released version? If so, the problem is most certainly #374. If not, it's definitely not #374.

**Comment by jrosdahl** [comment]
@jcdutton: Any answers to the follow-up questions?
